From 970f19bdb55cd559e9ef97228d30fd52b20e39cb Mon Sep 17 00:00:00 2001
From: Thiago Macieira <thiago.macieira@nokia.com>
Date: Tue, 8 Dec 2009 20:28:45 +0100
Subject: [PATCH] Fix crash in QDBusPendingReply/QDBusReply in case of unconnected calls.

If we made calls on a QDBusConnection that isn't connected, the d
pointer is 0. Ensure we don't crash.

Task-number: QTBUG-6571
Reviewed-by: Bradley T. Hughes
---
 src/dbus/qdbuspendingcall.cpp                      |    2 +-
 src/dbus/qdbuspendingreply.h                       |    1 +
 .../qdbuspendingreply/tst_qdbuspendingreply.cpp    |   25 ++++++++++++++++++++
 tests/auto/qdbusreply/tst_qdbusreply.cpp           |   16 ++++++++++++
 4 files changed, 43 insertions(+), 1 deletions(-)

--- a/src/dbus/qdbuspendingcall.cpp
+++ b/src/dbus/qdbuspendingcall.cpp
@@ -310,7 +310,7 @@ QDBusPendingCall &QDBusPendingCall::oper
 
 bool QDBusPendingCall::isFinished() const
 {
-    return d && (d->replyMessage.type() != QDBusMessage::InvalidMessage);
+    return !d || (d->replyMessage.type() != QDBusMessage::InvalidMessage);
 }
 
 void QDBusPendingCall::waitForFinished()
--- a/src/dbus/qdbuspendingreply.h
+++ b/src/dbus/qdbuspendingreply.h
@@ -188,6 +188,7 @@ public:
 private:
     inline void calculateMetaTypes()
     {
+        if (!d) return;
         int typeIds[Count > 0 ? Count : 1]; // use at least one since zero-sized arrays aren't valid
         ForEach::fillMetaTypes(typeIds);
         setMetaTypes(Count, typeIds);
